### PR TITLE
[SPARK-17585][PySpark][Core] PySpark SparkContext.addFile supports adding files recursively

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?

Users would like to add a directory as dependency in some cases, they can use `SparkContext.addFile` with argument `recursive=true` to recursively add all files under the directory by using Scala. But Python users can only add file not directory, we should also make it supported.
## How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by yanboliang** [comment]
cc @rxin @davies @srowen

File: python/pyspark/tests.py [hunk 1]
```diff
@@ -409,13 +409,22 @@ def func(x):
...
 
+        path = os.path.join(SPARK_HOME, "python/test_support/hello")
+        self.sc.addFile(path, True)
+        download_path = SparkFiles.get("hello")
+        self.assertNotEqual(path, download_path)
+        with open(download_path + "/hello.txt") as test_file:
+            self.assertEqual("Hello World!\n", test_file.readline())
+        with open(download_path + "/sub_hello/sub_hello.txt") as test_file:
+            self.assertEqual("Sub Hello World!\n", test_file.readline())
+
```
Comment (BryanCutler) [inline python/pyspark/tests.py]: minor:  maybe the above block should be in a separate test like `def test_add_file_locally_recursive`?

**Comment by BryanCutler** [comment]
Just one minor suggestion, otherwise LGTM

**Comment by yanboliang** [comment]
@BryanCutler Updated, thanks your comments.

**Comment by yanboliang** [comment]
@srowen Would you mind to have a look at this when you available? Thanks.

File: core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala [hunk 2]
```diff
@@ -670,6 +670,19 @@ class JavaSparkContext(val sc: SparkContext)
...
   /**
+   * Add a file to be downloaded with this Spark job on every node.
+   * The `path` passed can be either a local file, a file in HDFS (or other Hadoop-supported
+   * filesystems), or an HTTP, HTTPS or FTP URI.  To access the file in Spark jobs,
+   * use `SparkFiles.get(fileName)` to find its download location.
+   *
+   * A directory can be given if the recursive option is set to true. Currently directories are only
+   * supported for Hadoop-supported filesystems.
+   */
+  def addFile(path: String, recursive: Boolean): Unit = {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala]: (Why do we need this?)

File: python/pyspark/context.py [hunk 3]
```diff
@@ -762,7 +762,7 @@ def accumulator(self, value, accum_param=None):
         SparkContext._next_accum_id += 1
         return Accumulator(SparkContext._next_accum_id - 1, value, accum_param)
 
-    def addFile(self, path):
+    def addFile(self, path, recursive=False):
```
Comment (srowen) [inline python/pyspark/context.py]: This basically doesn't change the API right? you can still call it as before with the same behavior.

It seems reasonable to me overall because it adds parity between the APIs, isn't complex and doesn't change behavior.

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala (hunk 2 above)
Comment (yanboliang) [inline core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala]: Since ```JavaSparkContext``` is the Java stubs which will be called by PySpark.

File: python/pyspark/context.py (hunk 3 above)
Comment (yanboliang) [inline python/pyspark/context.py]: Yes, it does not change the existing API.

**Comment by yanboliang** [comment]
Merged into master. Thanks for all your review. @BryanCutler @srowen

File: core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala (hunk 2 above)
Comment (srowen) [inline core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala]: You're calling the method on `SparkContext`: 

```
self._jsc.sc().addFile(path, recursive)
```

I don't think this needed to be exposed?

Comment (yanboliang) [inline core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala]: Oh, I see. I found ```_jsc.sc()``` and ```_jsc``` are mix-used in ```context.py```. I will do some clean up and unify them in a follow up work. Thanks for your comments.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala]: Sounds good. There may be a reason the Java context is needed for some calls. I suppose that where the SparkContext could be used ... yeah that's simpler but doesn't really save anything because we wouldn't be able to take methods out of JavaSparkContext. That's why I was hoping to avoid adding a method to it.

Comment (yanboliang) [inline core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala]: @srowen After investigating the code, I found it's not very straightforward to clean up the interfaces at ```JavaSparkContext```, since they were called by Python and R. For Python side, we can use ```_jsc.sc()``` in some cases, but it's messy if we use both ```JavaSparkContext``` and ```JavaSparkContext.sc``` at R side. So I think we should leave it as it is, or any other suggestion? Thanks.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala]: Yes, but can we undo this change? it doesn't seem like we need to duplicate this method in the Java API.

Comment (yanboliang) [inline core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala]: I think it's not required to undo this, since I will send a PR to support recursively add files under a directory for SparkR soon and it will leverage this API. Thanks.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala]: OK, if that requires the Java context, I get it.
